Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -329,31 +329,47 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
-		//       stored in a single texture and uses the palette to make different characters
-		//       visible or invisible. Thus, unless we want to recreate the textures for every drawn character,
-		//       we must make sure that a paletted texture gets assigned multiple IDs for each tlut used.
-		//
-		//       EFB copys however didn't know anything about the tlut, so don't change the texID if there
-		//       already is an efb copy at this source. This makes those textures less broken when using efb to texture.
-		//       Examples are the mini map in Twilight Princess and objects on the targetting computer in Rogue Squadron 2(RS2).
-		// TODO: Convert those textures using the right palette, so they display correctly
-		auto iter = textures.find(texID);
-		if (iter == textures.end() || !iter->second->IsEfbCopy())
-			texID ^= ((u32)tlut_hash) ^(u32)(tlut_hash >> 32);
+		// TODO: Convert paletted textures, which are efb copies, using the right palette, so they display correctly
 	}
 
 	// GPUs don't like when the specified mipmap count would require more than one 1x1-sized LOD in the mipmap chain
 	// e.g. 64x64 with 7 LODs would have the mipmap chain 64x64,32x32,16x16,8x8,4x4,2x2,1x1,0x0, so we limit the mipmap count to 6 there
 	tex_levels = std::min<u32>(IntLog2(std::max(width, height)) + 1, tex_levels);
 
-	TCacheEntryBase*& entry = textures[texID];
+	TCacheEntryBase* entry = nullptr;
+
+	// Use multiple cache entries for textures at the same address, if they are loaded during the same frame. If this is not done, there is
+	// only one texture cache entry for those textures, which gets overwritten and recreated all the time, so effectively none of those
+	// textures is cached. Multiple cache entries prevent slowdowns in Metroid Prime and Castlevania 3. Metroid Prime has multiple sets of
+	// fonts on each other stored in a single texture and uses the palette to make different characters visible or invisible. In Castlevania 3
+	// all textures use the same address. If there were multiple cache entries for all textures at the same address, no texture cache entries
+	// would get overwritten at all. Overwriting a cache entry is faster than creating a new one, if it's unlikely that the old entry is used
+	// again. The restriction to textures loaded during the same frame, seems to be a good heuristic to determine when using multiple cache
+	// entries is faster and when not.
+	// For efb copies, the entry from CopyRenderTargetToTexture has to be used, or else it was done in vain.
+	std::pair <TexCache::iterator, TexCache::iterator> iter_range = textures.equal_range(texID);
+	TexCache::iterator iter = iter_range.first;
+	if (iter != iter_range.second)
+	{
+		if (iter->second->IsEfbCopy() || (iter->second->frameCount != FRAMECOUNT_INVALID && std::next(iter, 1) == iter_range.second))
+		{
+			entry = iter->second;
+		}
+		else
+		{
+			while (!entry && iter != iter_range.second)
```

Comment:
If I'm understanding this correctly, this is essentially an alternate mode: if we detect a game using multiple textures with the same address, we allow any number of textures with the same address.  This seems like it could easily blow up: if the heuristic gets triggered for one frame, it's essentially turned on permanently, which could kill performance.

Maybe just make Cleanup() a lot more aggressive in this case?